File: WoofAndroid/domain/usecase/src/main/kotlin/com/perrystreet/woof/usecase/favorites/GetFavoriteDogsUseCase.kt
package com.perrystreet.woof.usecase.favorites

import com.perrystreet.woof.models.dog.Dog
import com.perrystreet.woof.repositories.favorites.FavoritesRepository
import com.perrystreet.woof.usecase.dogs.GetDogsFeedUseCase
import io.reactivex.rxjava3.core.Observable
import org.koin.core.annotation.Factory

@Factory
class GetFavoriteDogsUseCase(
    private val getDogsFeedUseCase: GetDogsFeedUseCase,
    private val favoritesRepository: FavoritesRepository,
) {
    operator fun invoke(): Observable<List<Dog>> =
        Observable.combineLatest(
            getDogsFeedUseCase(),
            favoritesRepository.favoriteDogIds,
        ) { feed, favoriteIds ->
            feed.dogs.filter { it.id in favoriteIds }
        }.distinctUntilChanged()
}

File: WoofIOS/SwiftPackages/Domain/UseCase/Sources/UseCase/Favorites/GetFavoriteDogsUseCase.swift
import Combine
import DI
import Models
import Repositories

@Factory
public final class GetFavoriteDogsUseCase {
    private let getDogsFeedUseCase: GetDogsFeedUseCase
    private let favoritesRepository: FavoritesRepository

    public func callAsFunction() -> AnyPublisher<[Dog], Never> {
        Publishers.CombineLatest(
            getDogsFeedUseCase(),
            favoritesRepository.favoriteDogIds
        )
        .map { feed, favoriteIds in
            feed.dogs.filter { favoriteIds.contains($0.id) }
        }
        .removeDuplicates()
        .eraseToAnyPublisher()
    }
}
